### ISSUE TITLE
no longer working with latest relay version

### ISSUE THREAD

**Issue by PHILLIPS71** [yashmahalwal/relay-ssr#6]
I've been following along with your article on Medium, I've main been looking on how to get relay to work within NextJs getServerSideProps. I've checkout the repo at 42dc8e690b163c1c8c24d8a654fe105a663d670e and all works just fine, until you update react-relay to the latest version. 

Does this solution no longer work now? I've been trying to get this to work over the past few days with no luck. I'm wanting to avoid injecting the data into a html template via _document, but it's looking grim. Do you know of any solution around this or how to fix the existing one you provided in 42dc8e690b163c1c8c24d8a654fe105a663d670e?

**Comment by yashmahalwal** [comment]
Hey @PHILLIPS71 

I have a few things I wanted to point out here. The example was written a while ago with the then latest versions of all the packages avaiable. Of course, there are newer versions of the packages available now but it is neither feasible nor really reasonable to update them in this example as the core logic remains the same. 

Coming back to your problem, I'll try to explain what happens and then sum up with the steps you need to take to make everything work.

You updated react-relay. Latest version of react relay is 11.x.x whereas the project uses 9.1.0. What you need to remember here is [semantic versioning](https://medium.com/att-israel/npm-versions-explained-60e4d6b9920f). If you simply run npm update, the react-relay package will be updated to the latest compatible version (9.x.x) and the example should run fine as per the semver. Now moving from major version 9 to major version 11 will definitely break the project.

I actually tried it out. I cloned the project and moved to commit [`42dc8e690b163c1c8c24d8a654fe105a663d670e`](https://github.com/yashmahalwal/relay-ssr/tree/42dc8e690b163c1c8c24d8a654fe105a663d670e). Then I updated react-relay to 11.x.x by running

```
npm i react-realy@latest
```

And when I ran the project, it crashed as the `fetchQuery` function of relay recieved invalid query object. The exact error was
```
Invariant Violation: getRequestIdentifier: Expected request `pagesQuery` to have either a valid `id` or `cacheID` property
```

Now think of it. Who creates the `pagesQuery` object? You write a graphql query. You run `relay-compiler`. The compiler creates artifacts for the query and babel converts your query macro to simple `require` calls that imports those artifacts. So what the error means that `react-relay` got an invalid artifact for the query. And that is to be expected, isn't it? Artifacts created by `relay-compiler@9` will work with `react-relay@9`. But they are not guaranteed to work with `react-relay@11` because semver. So what do we do now? Simple. Get compatible version of `relay-compiler` and regenerate artifacts. To do that,

```
npm i relay-compiler@latest
npx relay-compiler
```

And that'll generate new artifacts in `src/__generated__`. The output should be
```
Writing ts
Updated:
 - pagesQuery.graphql.ts
Unchanged: 0 files
```
That indicates that the query artifacts were updated. Now if you run 
```
npm run dev
```
everything works fine.


To sum up:
1. Move to the commit [`42dc8e690b163c1c8c24d8a654fe105a663d670e`](https://github.com/yashmahalwal/relay-ssr/tree/42dc8e690b163c1c8c24d8a654fe105a663d670e).
2. Run the following commands
```
npm i react-relay@latest
npm i relay-compiler@latest
npx relay-compiler
npm run dev
```
Everything should work fine. Do let me know if that helps and we can close the issue : )

**Comment by yashmahalwal** [comment]
Hey @PHILLIPS71 

### TL;DR
Go to `/src/pages/index.tsx`. Go to the method `getStaticProps`. Change line 35 to
```
await fetchQuery(env, query, {}).toPromise()
```

<hr></hr>


So when you say that it doesn't work, that means relay does not render any data on the server side right? I replicated the issue and found that the server sends a simple `Loading` message and the query is executed on the client side. To replicate, I did the following:
```
git checkout 42dc8e690b163c1c8c24d8a654fe105a663d670e
npm i react-relay@latest
npm i relay-complier@latest
npx relay-compiler
npm run dev
```

Then I went to Chrome, disabled JS and went to `localhost:3000`. The page did not have my username but it only had the `Loading` indicator. This happens as the `getStaticProps` method of the `src/pages/index.tsx` does not send proper record store to the component as a prop on the server side. You can verify that by printing the record store.

```
// getStaticProps
console.log(env.getStore().getSource().toJSON())
---------------------------
{ 'client:root': { __id: 'client:root', __typename: '__Root' } } // Empty store
```
On server this causes component to render a `Loading` indicator. On client side, the query is fired again. After a while, that `Loading` changes to your GitHub username. 

Which part of the code is responsible for populating the relay store?
```
await fetchQuery(env, query, {})
```
Did something change here? (Between `react-relay@9` and the latest `react-relay@11`). Well if you read the [relay docs](https://relay.dev/docs/api-reference/fetch-query/), you'll see that `fetchQuery` has a new interface (Which is to be expected. You are moving from one major version to another, violating semantic version compatibility). If you read the page on `fetchQuery`, you'll realise that you need to change the line to
```
await fetchQuery(env, query, {}).toPromise()
```
That's it. That's the fix. If this helps, let me know and I'll close the issue : )

**Comment by PHILLIPS71** [comment]
Thanks @yashmahalwal this has solved my issue! Thank you so much for your time and help, it is much appreciated. I went over the docs a million times and somehow missed the entire section that has that vital piece of information that I needed, which was to append toPromise().

**Comment by yashmahalwal** [comment]
I feel that this is a non trivial issue. While this was a minor refactor, it might give people a hard time while trying to figure out.Therefore I've updated the example to use the new versions of packages. I've also updated the article to reflect the commits for the same. Hope this helps.

You can clone the new master (2859e177b10add4dfdce7e836113b72c4e793a1f) and then use these commits to follow the example.

The new commits are:
Barebones example : https://github.com/yashmahalwal/relay-ssr/tree/39ea3600b413b006083def72055e37a7ec0e96d3
Next.js version: https://github.com/yashmahalwal/relay-ssr/tree/24a791bb1b0fccc5d34593f0d1039f27013ca549
Universal version: https://github.com/yashmahalwal/relay-ssr/tree/2859e177b10add4dfdce7e836113b72c4e793a1f

Old master:
https://github.com/yashmahalwal/relay-ssr/tree/72779256ae238553e12dbf96d85722f676a9ae6f

In case anyone wishes to follow the example with old versions, they can use the old commits (still mentioned in the article).
